File: 프로그래머스/3/87694. 아이템 줍기/아이템 줍기.java
import java.util.*;

class Solution {
    static int[][] directions = {{1, 0}, {-1, 0}, {0, 1}, {0, -1}};
    
    public int solution(int[][] rectangle, int characterX, int characterY, int itemX, int itemY) {
        //맵의 크기 두배로 늘리기
        int maxCoord = 102;
        boolean[][] map = new boolean[maxCoord][maxCoord];
        
        // 직사각형 영역을 표시하고, 내부 공간을 비웁니다.
        for (int[] rect : rectangle) {
            int x1 = rect[0] * 2, y1 = rect[1] * 2, x2 = rect[2] * 2, y2 = rect[3] * 2;
            for (int i = x1; i <= x2; i++) {
                for (int j = y1; j <= y2; j++) {
                    map[i][j] = true;
                }
            }
        }
        
        // 기존 직사각형에서 테두리만 남기고 내부 공간 false 처리로 비우기
        for (int[] rect : rectangle) {
            int x1 = rect[0] * 2 + 1, y1 = rect[1] * 2 + 1, x2 = rect[2] * 2 - 1, y2 = rect[3] * 2 - 1;
            for (int i = x1; i <= x2; i++) {
                for (int j = y1; j <= y2; j++) {
                    map[i][j] = false;
                }
            }
        }
        
        return bfs(map, characterX * 2, characterY * 2, itemX * 2, itemY * 2);
    }//end of solution
    
    private int bfs(boolean[][] map, int startX, int startY, int endX, int endY) {
        int maxCoord = 102;
        boolean[][] visited = new boolean[maxCoord][maxCoord];
        Queue<int[]> queue = new LinkedList<>();
        
        queue.offer(new int[]{startX, startY, 0});
        visited[startX][startY] = true;
        
        while (!queue.isEmpty()) {
            int[] current = queue.poll();
            int x = current[0], y = current[1], dist = current[2];
            
            //도착지점에 도착하면 값을 반으로 나눠서 출력
            if (x == endX && y == endY) {
                return dist / 2;
            }
            
            //4방향으로 탐색하기
            for (int[] dir : directions) {
                int newX = x + dir[0];
                int newY = y + dir[1];
                
                //경계조건 만족하는지 체크하기
                if (newX >= 0 && newX < maxCoord && newY >= 0 && newY < maxCoord 
                    && map[newX][newY] && !visited[newX][newY]) {
                    visited[newX][newY] = true;
                    queue.offer(new int[]{newX, newY, dist + 1});
                }
            }
        }
        return -1;
    }//end of bfs
}

File: 프로그래머스/3/87694. 아이템 줍기/아이템 줍기.py
from collections import deque

def solution(rectangle, characterX, characterY, itemX, itemY):
    answer = 0
    board = [[0]*101 for _ in range(101)]
    characterX, characterY, itemX, itemY = characterX*2, characterY*2, itemX*2, itemY*2
    
    for x1, y1, x2, y2 in rectangle:
        x1, y1, x2, y2 = x1*2, y1*2, x2*2, y2*2
        
        for x in range(x1, x2+1):
            board[x][y1] = 1
            board[x][y2] = 1
            
        for y in range(y1, y2+1):
            board[x1][y] = 1
            board[x2][y] = 1
            
    for x1, y1, x2, y2 in rectangle:
        x1, y1, x2, y2 = x1*2, y1*2, x2*2, y2*2
        
        # 변 표시 후 내부를 0으로 지움
        for x in range(x1+1, x2):
            for y in range(y1+1, y2):
                board[x][y] = 0  # 내부는 이동 불가
                
                
    # print(board)
    q = deque([(characterX, characterY,1)])
    board[characterX][characterY] = 0
    dx = [0,0,1,-1]
    dy = [1,-1,0,0]
    while q:
        x, y, dist = q.popleft()
        if x == itemX and y == itemY:
            return dist//2
        
        for i in range(4):
            nx = x + dx[i]
            ny = y + dy[i]
            
            if 0<=nx<len(board) and 0<=ny<len(board[0]) and board[nx][ny] == 1:
                board[nx][ny] = 0 # 방문처리
                q.append((nx,ny,dist+1))
    
    # return answer